l5kit/l5kit/kinematic/ackerman_perturbation.py

```diff
@@ -4,162 +4,108 @@
-        return np.array([0.0, 0.0], dtype=np.float32)
-
-    point_to_be_perturbed = input_translations[perturbation_idx, :]
-    # we use this to find the local lateral direction
-    next_point_in_trajectory = input_translations[perturbation_idx + 1, :]
-
-    traj_dir_at_perturbation_point = next_point_in_trajectory - point_to_be_perturbed
-    #  the minimum distance between two consecutive trajectory points to go forward with direction computing.
-    consecutive_point_distance_threshold = 0.00001
-    #  the trajectory in the perturbation point may be a stationary point, so there's no direction info.
-    #  in this case, we just look at the start and end of the array to get the overall motion direction.
-    #  if that fails, we just don't perturb.
-
-    if np.linalg.norm(traj_dir_at_perturbation_point) < consecutive_point_distance_threshold:
-        traj_dir_at_perturbation_point = input_translations[-1, :] - input_translations[0, :]
-        if np.linalg.norm(traj_dir_at_perturbation_point) < consecutive_point_distance_threshold:
-            traj_dir_at_perturbation_point = np.array([0.0, 0.0], dtype=np.float32)
-        else:
-            traj_dir_at_perturbation_point /= np.linalg.norm(traj_dir_at_perturbation_point)
-    else:
-        traj_dir_at_perturbation_point /= np.linalg.norm(traj_dir_at_perturbation_point)
-
-    offset_dir = np.array([traj_dir_at_perturbation_point[1], -traj_dir_at_perturbation_point[0]])
-
-    return offset_distance * offset_dir
-
-
-def _get_history_and_future_frames_as_joint_trajectory(
-        history_frames: np.ndarray, future_frames: np.ndarray
-) -> np.ndarray:
-    num_history_frames = len(history_frames)
-    num_future_frames = len(future_frames)
-    total_trajectory_length = num_history_frames + num_future_frames
-
-    combined_trajectory = np.zeros((total_trajectory_length, 3), dtype=np.float32)
+def _get_trajectory(history_frames: np.ndarray, future_frames: np.ndarray) -> np.ndarray:
+    num_frames = len(history_frames) + len(future_frames)
+    trajectory = np.zeros((num_frames, 3), dtype=np.float32)
 
     # Note that history frames go backward in time from the anchor frame.
-    combined_trajectory[:num_history_frames, :2] = history_frames["ego_translation"][::-1, :2]
-    combined_trajectory[:num_history_frames, 2] = [
-        rotation33_as_yaw(rot) for rot in history_frames["ego_rotation"][::-1]
-    ]
-
-    combined_trajectory[num_history_frames:, :2] = future_frames["ego_translation"][:, :2]
-    combined_trajectory[num_history_frames:, 2] = [rotation33_as_yaw(rot) for rot in future_frames["ego_rotation"]]
-
-    return combined_trajectory
+    trajectory[:, :2] = np.concatenate(
+        (history_frames["ego_translation"][::-1, :2], future_frames["ego_translation"][:, :2])
+    )
+    rotations = np.concatenate((history_frames["ego_rotation"][::-1], future_frames["ego_rotation"]))
+    trajectory[:, 2] = [rotation33_as_yaw(rot) for rot in rotations]
```

Comment:
this is quite expensive, but I don't see way around it

Reply:
we could use rotation matrix directly instead of converting to yaw, but I will leave it for now